networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -372,7 +371,290 @@ def asadpour_tsp(G, weight="weight"):
+
+        min_in_edge = min(G.in_edges(n, data=True), key=lambda x: x[2][weight])
+        min_root["weight"] = min_root["weight"] + min_in_edge[2][weight]
+        max_root["weight"] = max_root["weight"] + min_in_edge[2][weight]
+
+        min_arb_weight = math.inf
+        for arb in nx.ArborescenceIterator(G_1):
+            # Only iterate over the minimum weight arborescences
+            arb_weight = arb.size(weight)
+            if min_arb_weight == math.inf:
+                min_arb_weight = arb_weight
+            elif arb_weight > min_arb_weight + max_root["weight"] - min_root["weight"]:
+                break
+            # We have to pick the root node of the arborescence for the out
+            # edge of the first vertex as that is the only node without an
+            # edge directed into it.
+            for N in arb:
+                if arb.in_degree(N) == 0:
+                    # root found
+                    arb.add_edge(n, N, weight=G[n][N][weight])
+                    arb_weight += G[n][N][weight]
+                    break
+            # We can pick the minimum weight in-edge for the vertex with a
+            # cycle
+            min_in_edge_weight = math.inf
+            min_in_edge = None
+            for u, v, d in G.in_edges(n, data=True):
+                edge_weight = d[weight]
+                if u == N:
+                    continue
+                if edge_weight < min_in_edge_weight:
+                    min_in_edge_weight = edge_weight
+                    min_in_edge = (u, v)
+            arb.add_edge(min_in_edge[0], min_in_edge[1], weight=min_in_edge_weight)
+            arb_weight += min_in_edge_weight
+            # Check to see the weight of the arborescence, if it is a new
+            # minimum, clear all of the old potential minimum
+            # 1-arborescences and add this is the only one. If its weight is
+            # above the known minimum, do not add it.
+            if arb_weight < minimum_1_arborescence_weight:
+                minimum_1_arborescences.clear()
+                minimum_1_arborescence_weight = arb_weight
+            # We have a 1-arborescence, add it to the set
+            if arb_weight == minimum_1_arborescence_weight:
+                minimum_1_arborescences.add(arb)
+
+        return minimum_1_arborescences
+
+    def direction_of_ascent():
+        """
+        Find the direction of ascent at point pi.
+
+        See [1]_ for more information.
+
+        Returns
+        -------
+        dict
+            A mapping from the nodes of the graph which represents the direction
+            of ascent.
+
+        References
+        ----------
+        .. [1] M. Held, R. M. Karp, The traveling-salesman problem and minimum
+           spanning trees, Operations Research, 1970-11-01, Vol. 18 (6),
+           pp.1138-1162
+        """
+        # 1. Set d equal to the zero n-vector.
+        d = {}
+        for n in G:
+            d[n] = 0
```

Comment:
Advanced performance suggestion; feel free to ignore (especially as I'm unsure of the performance impact here).  If you relabel the graph with `nx.relabel.convert_node_labels_to_integers` at the top level (and translate the solution back), you can use a list or numpy array (with `dtype`) instead of this dict to improve performance.